src/atomics/seqcst.md

```diff
@@ -0,0 +1,432 @@
+
+// Set this to Relaxed, Acquire, Release, AcqRel, doesn’t matter — the result is
+// the same (modulo panics caused by attempting acquire stores or release
+// loads).
+const ORDERING: atomic::Ordering = atomic::Ordering::Relaxed;
+
+static X: AtomicBool = AtomicBool::new(false);
+static Y: AtomicBool = AtomicBool::new(false);
+
+let a = thread::spawn(|| { X.store(true, ORDERING); Y.load(ORDERING) });
+let b = thread::spawn(|| { Y.store(true, ORDERING); X.load(ORDERING) });
+
+let a = a.join().unwrap();
+let b = b.join().unwrap();
+
+# return;
+// This assert is allowed to fail.
+assert!(a || b);
+```
+
+The basic setup of this code, for all of its possible executions, looks like
+this:
+
+```text
+     a        static X    static Y         b
+╭─────────╮   ┌───────┐   ┌───────┐   ╭─────────╮
+│ store X ├─┐ │ false │   │ false │ ┌─┤ store Y │
+╰────╥────╯ │ └───────┘   └───────┘ │ ╰────╥────╯
+╭────⇓────╮ └─┬───────┐   ┌───────┬─┘ ╭────⇓────╮
+│ load Y  ├─? │ true  │   │ true  │ ?─┤ load X  │
+╰─────────╯   └───────┘   └───────┘   ╰─────────╯
+```
+
+In other words, `a` and `b` are guaranteed to store `true` into `X` and `Y`
+respectively, and then attempt to load from the other thread’s atomic. The
+question now is: is it possible for them _both_ to load `false`?
+
+And looking at this diagram, there’s absolutely no reason why not. There isn’t
+even a single arrow connecting the left and right hand sides so far, so the
+loads have no coherence-based restrictions on which values they are allowed to
+pick, and we could end up with an execution like this:
+
+```text
+     a        static X    static Y         b
+╭─────────╮   ┌───────┐   ┌───────┐   ╭─────────╮
+│ store X ├┐  │ false ├─┐┌┤ false │  ┌┤ store Y │
+╰────╥────╯│  └───────┘┌─┘└───────┘  │╰────╥────╯
+     ║     │ ┌─────────┘└───────────┐│     ║
+╭────⇓────╮└─│┬───────┐   ┌───────┬─│┘╭────⇓────╮
+│ load Y  ├──┘│ true  │   │ true  │ └─┤ load X  │
+╰─────────╯   └───────┘   └───────┘   ╰─────────╯
+```
+
+Which results in a failed assert. This execution is brought about because the
+model of separate modification orders means that there is no relative ordering
+between `X` and `Y` being changed, and so each thread is allowed to “see” either
+order. However, some algorithms will require a globally agreed-upon ordering,
+and this is where `SeqCst` can come in useful.
+
+This ordering, first and foremost, inherits the guarantees from all the other
+orderings — it is an acquire operation for loads, a release operation for stores
+and an acquire-release operation for RMWs. In addition to this, it gives some
+guarantees unique to `SeqCst` about what values it is allowed to load. Note that
+these guarantees are not about preventing data races: unless you have some
+unrelated code that triggers a data race given an unexpected condition, using
+`SeqCst` can only prevent you from race conditions because its guarantees only
+apply to other `SeqCst` operations rather than all data accesses.
```

Comment:
I wonder whether it would be good to add at this point a note that despite popular belief that seqcst is the "strongest" ordering, it doesn't mean that every SeqCst operation is also Acquire or Release operation - SeqCst load cannot provide Release guarantee and vice versa.